bokeh/charts/_builder.py

```diff
@@ -0,0 +1,278 @@
+
+#-----------------------------------------------------------------------------
+# Classes and functions
+#-----------------------------------------------------------------------------
+
+def create_and_build(builder_class, values, **kws):
+    # create the new builder
+    builder = builder_class(values, **kws)
+
+    # create a chart to return, since there isn't one already
+    chart = Chart(**kws)
+    chart.add_builder(builder)
+
+    return chart
+
+
+class Builder(object):
+    """A prototype class to inherit each new chart type.
+
+    It provides useful methods to be used by the inherited chart classes,
+    such as:
+        * chained methods
+        * dynamically chart cration attributes, ie. `self.chart`
+        * composition to easily use Chart clase methods
+    """
+    def __init__(self, legend=False, palette=None, **kws):
+        """Common arguments to be used by all the inherited classes.
+
+        Args:
+            legend (str, bool): the legend of your plot. The legend content is
+                inferred from incoming input.It can be ``top_left``,
+                ``top_right``, ``bottom_left``, ``bottom_right``.
+                It is ``top_right`` is you set it as True.
+            palette(list, optional): a list containing the colormap as hex values.
+
+        """
+        self._legend = legend
+        self._palette = palette or DEFAULT_PALETTE
+        self._legends = []
+        self.data = {}
+        self.groups = []
+        self.attr = []
+        self.groups = []
+
+    def prepare_values(self):
+        """Prepare the input data.
+
+        Converts data input (self.values) to a DataAdapter and creates
+        instance index if needed
+        """
+        if hasattr(self, 'index'):
+            self.values_index, self.values = DataAdapter.get_index_and_data(
+                self.values, self.index
+            )
+        else:
+            if not isinstance(self.values, DataAdapter):
+                self.values = DataAdapter(self.values, force_alias=False)
+
+    def get_data(self):
+        """Get the input data.
+
+        It has to be implemented by any of the inherited class
+        representing each different chart type. It is the place
+        where we make specific calculations for each chart.
+        """
+        pass
+
+    def get_source(self):
+        """Push data into the ColumnDataSource and build the proper ranges.
+
+        It has to be implemented by any of the inherited class
+        representing each different chart type.
+        """
+        pass
+
+    def draw(self):
+        """Draw the glyphs into the plot.
+
+        It has to be implemented by any of the inherited class
+        representing each different chart type.
+        """
+        pass
+
+    def _pre_create(self):
+        """ Hook method that can be overwritten to inject chart logic to
+        be executed before make renderers method"""
+
```

Comment:
missing `pass`, I know we had this discussion before... about the no need if there is docstrings... but all the other methods have the `pass`, so let's be consistent...